core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -466,22 +466,25 @@ public final void runInContext() {
     }
 
     @Override
-    public void messageRead(final InputStream message) {
+    public void messagesAvailable(final MessageProducer producer) {
       class MessageRead extends ContextRunnable {
         MessageRead() {
           super(context);
         }
 
         @Override
         public final void runInContext() {
+          InputStream message;
           try {
-            if (closed) {
-              return;
-            }
-            try {
-              observer.onMessage(method.parseResponse(message));
-            } finally {
-              message.close();
+            while ((message = producer.next()) != null) {
+              if (closed) {
+                return;
+              }
+              try {
+                observer.onMessage(method.parseResponse(message));
+              } finally {
```

Comment:
Can you elaborate on what you're looking for here? This can fail to close some messages, if `producer.next()` throws an exception, but not sure what we should do about that here since the messages might not exist or have been deframed at that point. When the producer is a wrapper around a single message, as in this commit, isn't the try/finally here exactly the same as with the current code?